File: src/boj/boj_14888.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_14888 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] num = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            num[i] = Integer.parseInt(st.nextToken());
        }

        int[] arr = new int[n - 1];
        int idx = 0;
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) {
            int cnt = Integer.parseInt(st.nextToken());
            for (int j = 0; j < cnt; j++) {
                arr[idx++] = i;
            }
        }

        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;

        do {
            int value = num[0];
            for (int i = 1; i < n; i++) {
                switch (arr[i - 1]) {
                    case 0:
                        value += num[i];
                        break;
                    case 1:
                        value -= num[i];
                        break;
                    case 2:
                        value *= num[i];
                        break;
                    case 3:
                        value /= num[i];
                        break;
                }
            }
            max = Math.max(max, value);
            min = Math.min(min, value);
        } while (nextPermutation(arr));

        System.out.printf("%d\n%d\n", max, min);
    }
    static boolean nextPermutation(int[] arr) {
        int left = -1;
        for (int i = arr.length - 2; i >= 0; i--) {
            if (arr[i] < arr[i + 1]) {
                left = i;
                break;
            }
        }

        if (left == -1) {
            return false;
        }

        int right = left + 1;
        for (int i = arr.length - 1; i >= right; i--) {
            if (arr[left] < arr[i]) {
                right = i;
                break;
            }
        }

        int temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;

        for (int i = left + 1; i <= (arr.length + left) / 2; i++) {
            temp = arr[i];
            arr[i] = arr[arr.length + left - i];
            arr[arr.length + left - i] = temp;
        }
        return true;
    }
}

File: src/boj/boj_14888.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max
import kotlin.math.min

private var n = 0
private lateinit var numArray: IntArray
private lateinit var expArray: IntArray
private lateinit var curExp: IntArray
private var max = -1000000000
private var min = 1000000000

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    n = readLine().toInt()
    with(StringTokenizer(readLine())) {
        numArray = IntArray(n) {
            nextToken().toInt()
        }
    }
    with(StringTokenizer(readLine())) {
        expArray = IntArray(4) {
            nextToken().toInt()
        }
    }
    curExp = IntArray(n - 1)
    bt(0)

    println("$max\n$min")
    close()
}

private fun bt(depth: Int) {
    if (depth == n - 1) {
        val sum = calculate()
        max = max(max, sum)
        min = min(min, sum)
        return
    }

    for (i in 0 until 4) {
        if (expArray[i] > 0) {
            expArray[i] -= 1
            curExp[depth] = i
            bt(depth + 1)
            expArray[i] += 1
        }
    }
}

private fun calculate(): Int {
    var sum = numArray.first()
    curExp.forEachIndexed { index, i ->
        when (i) {
            0 -> sum += numArray[index + 1]
            1 -> sum -= numArray[index + 1]
            2 -> sum *= numArray[index + 1]
            3 -> sum /= numArray[index + 1]
        }
    }
    return sum
}